codalab/model/bundle_model.py

```diff
@@ -687,6 +693,64 @@ def remove_data_hash_references(self, uuids):
         with self.engine.begin() as connection:
             connection.execute(cl_bundle.update().where(cl_bundle.c.uuid.in_(uuids)).values({'data_hash': None}))
 
+    def update_bundle_contents_index(self, uuid, index):
+        rows = []
+        def recurse(entry, parent_path):
+            entry['bundle_uuid'] = uuid
+            entry['path'] = os.path.join(parent_path, entry['name'])
+            del entry['name']
+            if 'contents' in entry:
+                for child_entry in entry['contents']:
+                    recurse(child_entry, entry['path'])
+                del entry['contents']
+            if 'link' not in entry:
+                entry['link'] = None
+            rows.append(entry)
+        index_copy = copy.deepcopy(index)
+        if index_copy['name'] != uuid:
+            raise UsageError(
+                'Valid indices have the directory with the name as the UUID '
+                'being the top-level directory.')
+        # Clear the top-level name to save space in the database.
+        index_copy['name'] = ''
+        recurse(index_copy, '/')
+        with self.engine.begin() as connection:
+            connection.execute(cl_bundle_contents_index.delete().where(
```

Comment:
I'm a bit worried about bundles that have lots and lots of files...this might get pretty slow...
We don't have a cap on the number of files (maybe we should at least for intermediate updates).  In the past, a user has created 100K files.